### PR TITLE
JVM IR: Turn static callable references into singletons

### PR DESCRIPTION
**Pull request by sfs**
This is an optimization which is present in the JVM backend and in principle results in some observable differences in behavior (see added tests).

### CODE REVIEW

File: compiler/testData/codegen/box/coroutines/tailCallOptimizations/unit/override5_ir.txt [hunk 1]
```diff
@@ -3,19 +3,10 @@ public interface Base {
...
-    public method <init>(): void
-    public synthetic bridge @org.jetbrains.annotations.Nullable method invoke(): java.lang.Object
-    public final method invoke(): void
-}
-
 @kotlin.Metadata
 public final class Override5Kt$box$1$invokeSuspend$$inlined$inlineMe$1 {
     inner class Override5Kt$box$1$invokeSuspend$$inlined$inlineMe$1
-    inner class Override5Kt$box$1$invokeSuspend$$inlined$inlineMe$1$2
+    inner class Override5Kt$inlineMe$1$generic$2
```
Comment (dnpetrov) [inline compiler/testData/codegen/box/coroutines/tailCallOptimizations/unit/override5_ir.txt]: Minor: @ilmirus please, take a look.

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/StaticLambdaLowering.kt [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        val instanceField = backendContext.declarationFactory.getFieldForObjectInstance(constructor.constructedClass)
+        return IrGetFieldImpl(expression.startOffset, expression.endOffset, instanceField.symbol, expression.type)
+    }
+
+    // Recognize callable references with no value or type arguments. The only type arguments in Kotlin stem from usages of
+    // reified type parameters, which we unfortunately don't record as parameters so we have to check the body of the class.
+    private val IrClass.isSyntheticSingleton: Boolean
+        get() = (origin == JvmLoweredDeclarationOrigin.LAMBDA_IMPL || origin == JvmLoweredDeclarationOrigin.FUNCTION_REFERENCE_IMPL)
+                && primaryConstructor!!.valueParameters.isEmpty()
+                && !containsReifiedTypeParameters
```
Comment (dnpetrov) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/StaticLambdaLowering.kt]: I'd probably add some tests where this condition (no reified type parameters used) is relevant.

File: compiler/testData/codegen/box/coroutines/tailCallOptimizations/unit/override5_ir.txt (hunk 1 above)
Comment (ilmirus) [reply compiler/testData/codegen/box/coroutines/tailCallOptimizations/unit/override5_ir.txt]: LGTM

**Comment by dnpetrov** [comment]
Overall, looks good to me. I'll run tests and push manually.

**Comment by dnpetrov** [comment]
Pushed manually.

bb5a639153b483ef130ae2992fb2e73eba4cf5e6

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/StaticLambdaLowering.kt (hunk 2 above)
Comment (sfs) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/StaticLambdaLowering.kt]: There are already blackbox tests which test all of these conditions.

**Pull request closed without merging**